lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp

```diff
@@ -0,0 +1,472 @@
+           InstanceGraph &instanceGraph)
+      : module(module), combPathsMap(combPathsMap),
+        instanceGraph(instanceGraph) {}
+
+  /// Each operation visitor is supposed to finish the following tasks:
+  ///   1. Create a node for each of its results if the result is not a
+  ///   sequential value, such as register and memory write port;
+  ///   2. Create edges between its arguments and results if applicable.
+  using FIRRTLVisitor<Detector, bool>::visitStmt;
+  using FIRRTLVisitor<Detector, bool>::visitDecl;
+  using FIRRTLVisitor<Detector, bool>::visitExpr;
+
+  /// Statement handlers.
+  template <typename ConnectOpType>
+  bool handleConnectOp(ConnectOpType op);
+  bool visitStmt(ConnectOp op) { return handleConnectOp(op); }
+  bool visitStmt(PartialConnectOp op) { return handleConnectOp(op); }
+  bool visitStmt(WhenOp op) {
+    op->emitOpError("unsupported, please run -firrtl-expand-whens first");
+    return false;
+  }
+
+  /// Declaration handlers.
+  bool visitDecl(InstanceOp op);
+  bool visitDecl(MemOp op);
+  bool visitDecl(RegOp op) { return true; }
+  bool visitDecl(RegResetOp op) { return true; }
+
+  /// Expression handlers.
+  bool visitExpr(SubfieldOp op);
+  bool visitExpr(SubindexOp op) {
+    op->emitOpError("unsupported, please run -firrtl-lower-types first");
+    return false;
+  };
+  bool visitExpr(SubaccessOp op) {
+    op->emitOpError("unsupported, please run -firrtl-lower-types first");
+    return false;
+  };
+
+  /// Handle other operations that don't have concrete visitors.
+  bool visitUnhandledOp(Operation *op);
+
+  /// Build combinational graph of the module with the visitor dispatcher.
+  void buildCombGraph();
+
+  /// Detect combinational cycles in the module. Also detect combinational paths
+  /// between IOs of the module and record these paths in the `combPathsMap`.
+  void detect();
+
+private:
+  /// The module detected on.
+  FModuleOp module;
+
+  /// The global combinational paths map.
+  DenseMap<Operation *, CombPaths> &combPathsMap;
+
+  /// The global instance graph.
+  InstanceGraph &instanceGraph;
+
+  /// The local combinational graph.
+  CombGraph combGraph;
+};
+} // namespace
+
+template <typename ConnectOpType>
+bool Detector::handleConnectOp(ConnectOpType op) {
+  // Create an edge between the `source` value and `destination` value if both
+  // exist in the graph.
+  auto srcNode = combGraph.getNode(op.src());
+  auto dstNode = combGraph.getNode(op.dest());
+  if (srcNode && dstNode) {
+    auto connectNode = combGraph.getOrAddNode(op.getOperation());
+    srcNode->addChild(connectNode);
+    connectNode->addChild(dstNode);
+  }
+  return true;
+}
+
+/// Declaration handlers.
+bool Detector::visitDecl(InstanceOp op) {
+  SmallVector<CombNode, 8> nodes;
+  for (auto result : op.getResults())
+    nodes.push_back(combGraph.getOrAddNode(result));
+
+  // Query the combinational paths between IOs and create edges.
+  for (auto combPath :
```

Comment:
This is copying a combPath, i'd recommend referring to it by const& or something